### ISSUE TITLE
Choppy frame rate (probably due to Script/Phys) in VPW version of same table (Circus Voltaire)

### ISSUE THREAD

**Issue by WatchaGames** [vpinball/vpinball#2040]
Hi all,

First of all: Total respect for VPX devs, VPinMame devs and VPW crew.

Trying to figure out why the VPW tables are running choppy on my system (custom cabinet PC Based).
After much search/tests using Circus Voltaire as an example:
Looks like the visuals are not the problem: (tried with everything down like reallyyy down;-)

It looks like the VPW version of the table has jumps in both physics and script usage.
Here is a video showcasing the difference between the non VPW version and the VPW one.
[](https://youtube.com/shorts/aSwDoNPD2gI)
Circus Voltaire 1.1 and Circus Voltaire 1.1 VPW

As you can see there are big peaks in script/physics used time in frames.
Normal version runs 120hz, whereas VPW version looses frames.

I'm a bit puzzled by what could cause that.
Does anyone have an idea ?

Thx a lot.

PC Specs:(Should be ok) (many other tables run super well 120hz)
![PXL_20241204_231835002](https://github.com/user-attachments/assets/49e94753-10e2-41b7-9166-fee601a73321)
VPX: latest (RC8 and previous)
Looks like it is a bit less severe in latest RC8 version)
Windows: 10 (Pro removed all crap)

Context: I'm a game developer myself so I fancy being quite knowledgeable in 3d realtime apps.

**Comment by francisdb** [comment]
When you have the charts enabled vpinball should also log the slowest frames when you exit the table, there you can have a look at where time is spent.

```
2024-12-05 09:09:43.413 INFO  [504346] [FrameProfiler::LogWorstFrame@101] Long Frame of   1.8ms happened after  19.2s:
2024-12-05 09:09:43.413 INFO  [504346] [FrameProfiler::LogWorstFrame@177]   . Misc:             0.8ms
2024-12-05 09:09:43.413 INFO  [504346] [FrameProfiler::LogWorstFrame@177]   . Script:           0.3ms spent in   2 calls of TimerEvents:Timer (1 x RollingTimer =>  0.3ms, 1 x flippertimer =>  0.0ms) 
2024-12-05 09:09:43.413 INFO  [504346] [FrameProfiler::LogWorstFrame@177]   . Prepare Frame:    0.3ms
2024-12-05 09:09:43.413 INFO  [504346] [FrameProfiler::LogWorstFrame@177]   . Submit Frame:     0.2ms
2024-12-05 09:09:43.413 INFO  [504346] [FrameProfiler::LogWorstFrame@177]   . GPU Flip:         0.2ms
```

This was from a very low resources table (1000fps) but I currently have issues with vpinball not exiting causing these to not be shown for Circus Voltaire 1.1 VPW.

By the way, on full HD I have ~400 fps for CV VPW here. No hiccups. (Linux / OpenGL / Radeon RX 7600)

**Comment by WatchaGames** [comment]
Ok thx will run again and get that dump. brb

**Comment by francisdb** [comment]
Not sure if this could be related but there are some shader issues

```
2024-12-05 13:03:53.232 ERROR [12072] [Shader::getUniformByName@311] [FBShader.hlsl] getUniformByName Could not find uniform ms_zpd_ya_td in shaderUniformNames
2024-12-05 13:03:53.232 ERROR [12072] [Shader::Load@1551] Missing uniform: ms_zpd_ya_td
2024-12-05 13:03:53.236 ERROR [12072] [Shader::getUniformByName@311] [StereoShader.hlsl] getUniformByName Could not find uniform Stereo_Saturation in shaderUniformNames
2024-12-05 13:03:53.236 ERROR [12072] [Shader::Load@1551] Missing uniform: Stereo_Saturation
```

**Comment by francisdb** [comment]
Other than that I see no extreme frame times

**Comment by WatchaGames** [comment]
Will check if other VPW tables that I see behaving the same have this shader "problem".
Also will relog.. as I see quitge big peaks in the visual graph... will check why/how to find them in the log.

**Comment by WatchaGames** [comment]
ok made additional tests.
(Clean Install of latest VPX and VPinMame)
Tried with VpinMame 3.5 -> makes no diff.

Here is rerun -> looks like something to do with Physics ?
[vpinball.log](https://github.com/user-attachments/files/18049837/vpinball.log)

Also noticed, that when running the first time the table the spikes both in script and physics are much worse than on second run.

Sorry to not be more precise.
Hope it still helps.

**Comment by francisdb** [comment]
This is pretty crazy

> 2024-12-07 15:28:23.064 INFO  [13072] [FrameProfiler::LogWorstFrame@169]     . 1014.9ms spent in   1 calls of GameEvents:KeyUp

Can you play a bit longer to get some more long frames to validate they are all in the same place?

**Comment by WatchaGames** [comment]
Hi, thx for taking the time to look for a root "cause".
Makes sense as running scripts into scripts in modern OS became suspicious !
Will investigate security RTP on the machine to try to narrow down this.

**Comment by Crrispy** [comment]
That's strange. I just had a 43M full game (VPW), with extraball, several multiballs, it was 100% butter smooth at 4K120, and I'm very sensitive to stuttering, I hate it :-) 
 
Currently running the latest 10.8.0 build (2043), modest cpu (9700K @4.5Ghz) and 4070Ti.

**Comment by WatchaGames** [comment]
@Crrispy thx. yes it is  "strange", so far from what we observed/gathered, looks like the issue might be caused by something happening when executing a script from a script.
ie: as @francisdb mentioned:

VPinMame VBS script calls another script which might trigger OS/System realtime analysis of threats (calling/eval/executing) scripts from scripts is generaly seen as a dangerous thing by OSes/Browsers.

Tried to switch to OFF all kind of realtime protection tools/anti viruses with some success but not complete.

It is mostly happening with VPW tables.

@Crrispy might I ask what is your env ? OS ? Settings ?

Thx

**Comment by Crrispy** [comment]
Win 11 24H2, with Defender totally disabled via Defender Control 2.1. Quite standard settings, vsync/gsync off, frame pacing on, with max framerate to 120 in VPX.

**Comment by WatchaGames** [comment]
@Crrispy thx.
Defender Control 2.1 saved the day!
Somehow one thing was not disabled.
Will try to pinpoint which one and think about what could be done to avoid executing scripts withing scripts as this is generaly not a good idea.

**Comment by ZiggyDaMoe** [comment]
Where is a safe place to download Defender Control 2.1?

**Comment by WatchaGames** [comment]
> Where is a safe place to download Defender Control 2.1?

Not clear. I think I used MajorGeeks

**Comment by Crrispy** [comment]
> Where is a safe place to download Defender Control 2.1?

I got it from sordum, the developers.

**Comment by Drybonz** [comment]
If you are just having trouble with the one table, did you post about it on the comments where the table is posted?  It seems like you would get much more relevant feedback and troubleshooting advice.

**Comment by WatchaGames** [comment]
@Drybonz good point. Though looks like from the source code, that this is a "generic" "problem", about calling a VS script from a VS which triggers RT "virus/threat" windows systems.
Picked the biggest offender table as an illustration.

**Comment by Drybonz** [comment]
I guess if your anti-virus is false positive identifying a script as a threat that might cause some problem.  The easy way to find out would be to disable your anti virus and play some tables and see if it makes any difference.

**Comment by WatchaGames** [comment]
Yes thats we figured out. althrough ALL threat RT detection must be OFF.
Which is not ideal and if your Pinball PC Cabinet is connected to internet or if you play in desktop mode on your "main" Desk PC is even less than ideal.

**Comment by Drybonz** [comment]
Yeah, antivirus can be a pain.  I guess if your issue is with your AV then you can probably ask to have this issue closed.  Good luck with your problem.

**Comment by Loafmeister** [comment]
> Yes thats we figured out. althrough ALL threat RT detection must be OFF. Which is not ideal and if your Pinball PC Cabinet is connected to internet or if you play in desktop mode on your "main" Desk PC is even less than ideal.

It is common practice in the vpx/fp community to set your antivirus to exclude the affected folders. Perhaps that is sufficient to bypass this issue while also protecting your pc?  It’s how I have mine set up and no noticeable issues

**Comment by WatchaGames** [comment]
Yes I understand, thats the weird thing here, we tried excluding folders from anti virus scanning but only switching off completely threat defense would improve situation.
I know . does not make much sense why executing a VB script from a VS script which are in the same folder would still trigger.
Anyways, plan to re-install system soon from scratch and will see if I get the same result.
Thx

**Comment by WatchaGames** [comment]
Hi all, update with ALL latest releases (as of 2025-02-02)
Fresh Win 10 Install cleaned.
Reinstall of simple VPX setup thru the new installer.
Disabled all threat detection of Win + Excluded VPX folder from monitored folders (just in case disabling does not mean disabled in Windows ;-)

Still have big spikes VB code in this table.
BUT:
Noticed a big difference in frame rate stability and less/no spikes between:
CABINET (lots of spikes) vs Desktop (FSS or NoFSS) stable frame rate. (120)
Does not make much sense to me.
But what could the diff between CABINET and Destop mod which could have an impact on the way scripts called from scripts would impact perfs ?
Hope it heps.
